src/saturn_engine/worker/topics/delayed.py

```diff
@@ -0,0 +1,57 @@
+from typing import AsyncContextManager
+from typing import AsyncGenerator
+
+import asyncio
+import dataclasses
+from datetime import datetime
+from datetime import timedelta
+
+import asyncstdlib as alib
+
+from saturn_engine.core.topic import TopicMessage
+from saturn_engine.utils import utcnow
+from saturn_engine.utils.options import asdict
+from saturn_engine.utils.options import fromdict
+from saturn_engine.worker.services import Services
+from saturn_engine.worker.topics.rabbitmq import RabbitMQTopic
+
+
+@dataclasses.dataclass
+class DelayedTopicMetadata:
+    not_before: datetime
+
+
+class DelayedTopic(RabbitMQTopic):
+    @dataclasses.dataclass
+    class Options(RabbitMQTopic.Options):
+        delay: float = 60
+
+    def __init__(self, options: Options, services: Services, **kwargs: object) -> None:
+        super().__init__(options=options, services=services, **kwargs)
+        self.delay = options.delay
+
+    async def run(
+        self,
+    ) -> AsyncGenerator[AsyncContextManager[TopicMessage], None]:
+        async for message_context in super().run():
+            async with message_context as message:
```

Comment:
Not sure I understand the last example, but I think I can do this to avoid entering the message early:

```python
async def run_and_sleep() -> AsyncGenerator[AsyncContextManager[str], None]:
    async for message_context in run():
        yield with_and_sleep(message_context)

@asynccontextmanager
async def with_and_sleep(message_context: AsyncContextManager[str]) -> AsyncIterator[str]:
    async with message_context as value:
        print("sleep")
        yield value
```

Reply:
yeah work for me!

Reply:
Full example for funzies:

```python
import asyncio
from contextlib import asynccontextmanager, nullcontext
from typing import AsyncContextManager, AsyncGenerator, AsyncIterator


@asynccontextmanager
async def process() -> AsyncIterator[str]:
    print("enter message.process")
    try:
        yield "Hello World"
    finally:
        print("exit message.process")


@asynccontextmanager
async def message_context() -> AsyncIterator[str]:
    print("enter message_context")
    async with process() as value:
        yield value

    print("exit message_context")


async def run() -> AsyncGenerator[AsyncContextManager[str], None]:
    yield message_context()


async def run_and_sleep() -> AsyncGenerator[AsyncContextManager[str], None]:
    async for message_context in run():
        yield with_and_sleep(message_context)


@asynccontextmanager
async def with_and_sleep(message_context: AsyncContextManager[str]) -> AsyncIterator[str]:
    async with message_context as value:
        print("sleep")
        yield value


async def run_bad() -> AsyncGenerator[AsyncContextManager[str], None]:
    async for message_context in run():
        async with message_context as value:
            print("sleep")
            yield nullcontext(value)


async def main():
    print("--- Nothing should be printed here (context is never entered) ---")
    async for a in run_and_sleep():
        pass
    print("-" * 65)

    async for a in run_and_sleep():
        async with a as val:
            print(val)


asyncio.run(main())
```